CondCore/CondHDF5ESSource/plugins/HDF5ProductResolver.cc

```diff
@@ -0,0 +1,249 @@
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "h5_DataSet.h"
+#include "h5_Attribute.h"
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+HDF5ProductResolver::HDF5ProductResolver(edm::SerialTaskQueue* iQueue,
+                                         std::unique_ptr<cond::serialization::SerializationHelperBase> iHelper,
+                                         cms::h5::File const* iFile,
+                                         std::string const& iFileName,
+                                         cond::hdf5::Compression iCompression,
+                                         cond::hdf5::Record const* iRecord,
+                                         cond::hdf5::DataProduct const* iDataProduct)
+    : edm::eventsetup::ESSourceProductResolverBase(),
+      queue_(iQueue),
+      helper_(std::move(iHelper)),
+      file_(iFile),
+      fileName_(iFileName),
+      record_(iRecord),
+      dataProduct_(iDataProduct),
+      compression_(iCompression) {}
+
+// HDF5ProductResolver::HDF5ProductResolver(const HDF5ProductResolver& rhs)
+// {
+//    // do actual copying here;
+// }
+
+HDF5ProductResolver::~HDF5ProductResolver() {}
+
+//
+// member functions
+//
+
+void HDF5ProductResolver::prefetchAsyncImpl(edm::WaitingTaskHolder iTask,
+                                            edm::eventsetup::EventSetupRecordImpl const& iRecord,
+                                            edm::eventsetup::DataKey const& iKey,
+                                            edm::EventSetupImpl const*,
+                                            edm::ServiceToken const&,
+                                            edm::ESParentContext const& iParent) {
+  prefetchAsyncImplTemplate(
+      [this, iov = iRecord.validityInterval(), iParent, &iRecord](auto& iGroup, auto iActivity) {
+        queue_->push(iGroup, [this, &iGroup, act = std::move(iActivity), iov, iParent, &iRecord] {
+          try {
+            edm::ESModuleCallingContext context(
+                providerDescription(), edm::ESModuleCallingContext::State::kRunning, iParent);
+            iRecord.activityRegistry()->preESModuleSignal_.emit(iRecord.key(), context);
+            struct EndGuard {
+              EndGuard(edm::eventsetup::EventSetupRecordImpl const& iRecord,
+                       edm::ESModuleCallingContext const& iContext)
+                  : record_{iRecord}, context_{iContext} {}
+              ~EndGuard() { record_.activityRegistry()->postESModuleSignal_.emit(record_.key(), context_); }
+              edm::eventsetup::EventSetupRecordImpl const& record_;
+              edm::ESModuleCallingContext const& context_;
+            } guardAR(iRecord, context);
+
+            auto index = indexForInterval(iov);
+
+            readFromHDF5api(index);
+            iGroup.run(std::move(act));
+            exceptPtr_ = {};
+          } catch (...) {
```

Comment:
This "catch" is apparently caught (pun not intended 😄) as Error from the "CMSSW Code rules" [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d5b116/33993/codeRules/cmsCodeRule3.txt).

Reply:
The coding rules are meant to warn about `catch(...)` where the exception is just lost. But here the following lines grab the current exception and then properly pass it off to what ever thread will process the next 'task' in the chain. This is how we have to propagate exception in the multi-threaded system.

Reply:
ok! thanks!